deltachat-ffi/src/lib.rs

```diff
@@ -4424,3 +4425,77 @@ pub unsafe extern "C" fn dc_accounts_get_next_event(
         .map(|ev| Box::into_raw(Box::new(ev)))
         .unwrap_or_else(ptr::null_mut)
 }
+
+#[cfg(feature = "jsonrpc")]
+mod jsonrpc {
+    use super::*;
+    use deltachat_jsonrpc::api::CommandApi;
+    use deltachat_jsonrpc::yerpc::{MessageHandle, RpcHandle};
+
+    pub struct dc_json_api_instance_t {
+        receiver: async_std::channel::Receiver<deltachat_jsonrpc::yerpc::Message>,
+        handle: MessageHandle<CommandApi>,
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn dc_get_json_api(
+        account_manager: *mut dc_accounts_t,
+    ) -> *mut dc_json_api_instance_t {
+        if account_manager.is_null() {
+            eprintln!("ignoring careless call to dc_get_json_api()");
+            return ptr::null_mut();
+        }
+
+        let cmd_api =
+            deltachat_jsonrpc::api::CommandApi::new_from_cffi((*account_manager).inner.clone());
+
+        let (request_handle, receiver) = RpcHandle::new();
+        let handle = MessageHandle::new(request_handle, cmd_api);
+
+        let instance = dc_json_api_instance_t { receiver, handle };
+
+        Box::into_raw(Box::new(instance))
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn dc_json_api_unref(json_api_instance: *mut dc_json_api_instance_t) {
+        if json_api_instance.is_null() {
+            eprintln!("ignoring careless call to dc_json_api_unref()");
+            return;
+        }
+
+        Box::from_raw(json_api_instance);
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn dc_json_request(
+        json_api_instance: *mut dc_json_api_instance_t,
+        request: *const libc::c_char,
+    ) {
+        if json_api_instance.is_null() || request.is_null() {
+            eprintln!("ignoring careless call to dc_json_request()");
+            return;
+        }
+
+        let api = &*json_api_instance;
+        let handle = &api.handle;
+        let request = to_string_lossy(request);
+        async_std::task::spawn(async move {
+            handle.handle_message(&request).await;
+        });
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn dc_get_next_json_response(
+        json_api_instance: *mut dc_json_api_instance_t,
+    ) -> *mut libc::c_char {
+        if json_api_instance.is_null() {
+            eprintln!("ignoring careless call to dc_get_next_json_response()");
+            return ptr::null_mut();
+        }
+        let api = &*json_api_instance;
+        async_std::task::block_on(api.receiver.recv())
+            .map(|result| serde_json::to_string(&result).unwrap_or_default().strdup())
+            .unwrap_or(ptr::null_mut())
```

Comment:
I'm not sure if the error handling logic is correct here? Wondering whether the `unwrap_or_default()` makes sense - this would mean that if an invalid JSON value was returned (can that even happen?) that would be mapped to an empty string (which likely is not expected in code using that API) - so maybe better just `expect("Bad return value")` or default to the `ptr::null_mut()` and document that in the FFI API docs (which would mean that downstream code had to check for null pointers before using the strings - not sure if that is expected. I think I'd prefer to panic() in these cases, but not sure what the general error handling in deltachat CFFI is like)

Reply:
panic sounds bad, also `serde_json::to_string` should never fail in practice here